Add a separate record for edge elevation.

dnesbitt61 opened the pull request:
Moved max_grade (up and down) from directed edge to the new EdgeElevation record. Added a slot for mean_elevation (not yet computed and added in). If elevation data exists within a tile, the EdgeElevation count is the same as the DirectedEdge count - so edge elevation is accessed via directed edge GraphId. Much of the work in this PR is around maintaining the 1:1 correlation as the hierarchy is built and as shortcut edges are added. It was important to pad the tile data prior to adding lane connectivity so that the structures were aligned to a word boundary (padded to 8 bytes). Some sanity checks are made during tile building to make sure the end offset and size within the memory buffer are consistent and that edge elevation counts are consistent with directed edge counts as the tile is being built.

This was tested on a small OSM extract. Prior to merging a wider test on a larger extract will be performed.

@noblige - Please review and comment if you see anything that doesn't suit your needs. Next step is to add mean-elevation to skadi and GraphBuilder (and shortcut builder) where elevation is sampled.

noblige commented on src/thor/trippathbuilder.cc:
```diff
@@ -1260,13 +1260,17 @@ TripPath_Edge* TripPathBuilder::AddTripEdge(const TripPathController& controller
-  if (controller.attributes.at(kEdgeMaxDownwardGrade))
-    trip_edge->set_max_downward_grade(directededge->max_down_slope());
+  // Set maximum upward and downward grade if requested
+  if (controller.attributes.at(kEdgeMaxUpwardGrade) ||
```
need to add elevation here as well

dnesbitt61 replied:
latest addition added mean_elevation as well. Will add code to populate it within data to a separate PR.

noblige commented on valhalla/baldr/edge_elevation.h:
```diff
@@ -0,0 +1,84 @@
+// Use elevation bins of 8 meters to store mean elevation. Clamp to a range
+// from -500 meters to 7683 meters.
+constexpr uint32_t kMaxStoredElevation = 1023;
+constexpr float kElevationBinSize = 8.0f;
```
👍

noblige replied:
@dnesbitt61 Dave, actually can we increase resolution to 1m (13 bits), in case we want to change elevation data source (with a higher precision one).

Alternatively - move `mean_elevation_` down next to spare bits, so it can be changed afterwards

dnesbitt61 replied:
I will move it to just before spare, but that would still mean data incompatibility if we end up changing the precision at a future time. I can change to 13 (1m precision) if you want. I just wonder since this is "mean" elevation over the length of an entire edge I'm not sure how precise you really need. For now I'll change it to 12 bits (2m precision) and test. If you decide you want 1m then I will change to that.

noblige replied:
@dnesbitt61 I see your point - I was thinking more of a short, relatively flat segments. but 12 bits sounds good! Let's do that.

noblige commented on src/baldr/edge_elevation.cc:
```diff
@@ -0,0 +1,53 @@
+}
+
+// Set the mean elevation.
+void EdgeElevation::set_mean_elevation(const float mean_elev) {
```
thinking here what will happen if skadi returns `NO_DATA_VALUE` for sampled heights  (or if tiles are built without elevation data) - this gets then clamped to `kMinElevation`. wonder if it's better to introduce an explicit separate sentinel value (`kNoElevationData`) - this will be returned from `EdgeElevation::elevation()` when there's no data. That is:

`mean_elevation_ = 0` means `kNoElevationData`
`mean_elevation_ = 1` means `kMinElevation`
`kElevationNoData = -32768`

In other words something along the lines:

```
if (mean_elev > kMaxElevation || mean_elev < kMinElevation)
  mean_elevation_ = 0;
else
 mean_elevation_ = (mean_elev - (kMinElevation - 1)) / kElevationBinSize;
```

and for `elevation()` something like:
`return elevation_ > 0 ? ((mean_elevation_ * kElevationBinSize) + (kMinElevation - 1)) : kElevationNoData;`

Thoughts?

dnesbitt61 replied:
will look into this. GraphTile currently returns a pointer (or nullptr if the tile does not have elevation). Maybe have a static NoData EdgeElevation struct that gets pointed to in these cases.

dnesbitt61 commented:
@noblige: This looked fine with Maryland data. I added the mean elevation and looked at a route that I rode my bicycle on, so I know the elevation matches well. I also added some tests that helped validate the mean_elevation code. Will probably merge this on Monday unless there are remaining concerns.
